service/storage_service.cc

```diff
@@ -1441,7 +1441,7 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
         if (_sys_ks.local().bootstrap_complete()) {
```

Comment:
`storage_service: prepare_replacement_info: load ignored nodes state`

One thing I don't understand about this commit.

> But since the down nodes have unknown status, they never go
through `handle_state_normal` and therefore they are not accounted
as normal token owners.

But in previous commit, you changed that:
`storage_service: topology_state_load: set local STATUS state using add_saved_endpoint`

Wasn't the point of that commit precisely to make sure that such down nodes appear as NORMAL, therefore, when a node starts the replace operation, it will learn about all NORMAL nodes through `handle_state_normal` --- including the ignored nodes? So we shouldn't have to give special treatment to ignored nodes.

We should insert **all** normal nodes to token ring when we boot, ignored or not.

Nodes that are normal but not ignored, should also get inserted to token ring. If not, we will calculate replicas incorrectly, like in https://github.com/scylladb/scylla-enterprise/issues/3518. It's not really about ignored nodes -- it's about us missing tokens of some nodes.

So I guess my question is, why do we need this `storage_service: prepare_replacement_info: load ignored nodes state` commit -- doesn't the STATUS commit already fix the issue?

Reply:
@kbr-scylla:

> `storage_service: prepare_replacement_info: load ignored nodes state`
> 
> One thing I don't understand about this commit.
> 
> > But since the down nodes have unknown status, they never go
> > through `handle_state_normal` and therefore they are not accounted
> > as normal token owners.
> 
> But in previous commit, you changed that: `storage_service: topology_state_load: set local STATUS state using add_saved_endpoint`
> 
> Wasn't the point of that commit precisely to make sure that such down nodes appear as NORMAL, therefore, when a node starts the replace operation, it will learn about all NORMAL nodes through `handle_state_normal` --- including the ignored nodes? So we shouldn't have to give special treatment to ignored nodes.

This happens only for consistent-topology-changes in `storage_service::topology_state_load`, not with gossip.
For the latter we need this patch to load the ignored nodes state in prepare_replacement_info.

> We should insert **all** normal nodes to token ring when we boot, ignored or not.

I agree.
But the replacing node doesn't know about them in advance since it doesn't have system.peers yet.
My original patch loaded all normal token owners in add_saved_endpoint, setting their status initially to shutdown, which is considered a NORMAL status, giving them a chance to become be marked alive when we successfully communicate with them.  But Asias objected to that and preferred the endpoint state to have unknown STATUS - and the consequence is that it stays hidden in gossip - hence the need to add the saved state after a shadow round for replacing node.

> Nodes that are normal but not ignored, should also get inserted to token ring. If not, we will calculate replicas incorrectly, like in [scylladb/scylla-enterprise#3518](https://github.com/scylladb/scylla-enterprise/issues/3518). It's not really about ignored nodes -- it's about us missing tokens of some nodes.
> 
> So I guess my question is, why do we need this `storage_service: prepare_replacement_info: load ignored nodes state` commit -- doesn't the STATUS commit already fix the issue?

See above.

I feel like we're going in circles for this PR (that now needs to be rebased yet again).
It can be simplified with the approach I mentioned above - that is to always add the saved state as NORMAL/shutdown, if @asias removes his objection to that.
I agree with you that it would be safer this way, since is the best we know about those endpoints, given their state in system.peers (and since we don't save their last STATUS persistently)